Kernel/Interrupts/IOAPIC.cpp

```diff
@@ -227,50 +237,54 @@ bool IOAPIC::is_vector_enabled(u8 interrupt_vector) const
 
 u8 IOAPIC::read_redirection_entry_vector(u8 index) const
 {
-    ASSERT((u32)index < m_redirection_entries);
+    ASSERT((u32)index < m_redirection_entries_count);
     return (read_register((index << 1) + IOAPIC_REDIRECTION_ENTRY_OFFSET) & 0xFF);
 }
 
 int IOAPIC::find_redirection_entry_by_vector(u8 vector) const
 {
     InterruptDisabler disabler;
-    for (size_t index = 0; index < m_redirection_entries; index++) {
+    for (size_t index = 0; index < m_redirection_entries_count; index++) {
         if (read_redirection_entry_vector(index) == (InterruptManagement::acquire_mapped_interrupt_number(vector) + IRQ_VECTOR_BASE))
             return index;
     }
     return -1;
 }
 
-void IOAPIC::disable(u8 interrupt_vector)
+void IOAPIC::disable(GenericInterruptHandler& handler)
 {
     InterruptDisabler disabler;
     ASSERT(!is_hard_disabled());
-    int index = find_redirection_entry_by_vector(interrupt_vector);
+    ASSERT(handler.interrupt_number() >= gsi_base() && handler.interrupt_number() < interrupt_vectors_count());
+    int index = find_redirection_entry_by_vector(handler.interrupt_number());
     if (index == (-1)) {
-        map_interrupt_redirection(interrupt_vector);
-        index = find_redirection_entry_by_vector(interrupt_vector);
+        map_interrupt_redirection(handler.interrupt_number() - gsi_base());
+        index = find_redirection_entry_by_vector(handler.interrupt_number());
     }
     ASSERT(index != (-1));
     mask_redirection_entry(index);
 }
 
-void IOAPIC::enable(u8 interrupt_vector)
+void IOAPIC::enable(GenericInterruptHandler& handler)
 {
     InterruptDisabler disabler;
     ASSERT(!is_hard_disabled());
-    int index = find_redirection_entry_by_vector(interrupt_vector);
+    ASSERT(handler.interrupt_number() >= gsi_base() && handler.interrupt_number() < interrupt_vectors_count());
```

Comment:
This needs a conversion. Now it'll work but later when the mapping between IRQ lines to IDT vectors will be complicated, this will break.

Reply:
Or maybe not.... in `GenericInterruptHandler` we assign `m_interrupt_number` without any conversion.